### ISSUE TITLE
pylint exceed maximum recursion depth

### ISSUE THREAD

**Issue by skirpichev** [PyCQA/pylint#2388]
There are similar errors in bugtracker, e.g. #1410, #1285, but I'm not sure the underlying reason for this issue is same.

### Steps to reproduce
1. python3 -m venv ~/venv/pylint-test
2. . ~/venv/pylint-test/bin/activate
3. Follow http://diofant.readthedocs.io/en/latest/install.html#from-sources
4. run ``pylint --enable=old-raise-syntax diofant/solvers/recurr.py ``

### Current behavior

It fails with
```
internal error with sending report for module ['diofant/solvers/recurr.py']
maximum recursion depth exceeded
```

### Expected behavior

No internal error.

### pylint --version output

```
pylint 2.1.0
astroid 2.0.2
Python 3.5.3 (default, Jan 19 2017, 14:11:04) 
[GCC 6.3.0 20170118]
```

### Some debugging

Note, that same error happens if you enable different checker instead, almost any disabled.  For example, ``exception-message-attribute``.  The internal error disappears if I comment out [line 527](https://github.com/diofant/diofant/blob/b50066efea3974cb0249c6bed6cab3fae9dc37c2/diofant/solvers/recurr.py#L527).

**Comment by brycepg** [comment]
Here's my somewhat easier reproduction:

0. Install `diofant`
1. Save the following code in a file named `file.py`
```
from diofant.simplify import hypersimp, simplify

def rsolve_hyper():
    b = simplify(g)
    s = hypersimp(b, n)
    s.subs()
```
2. Run
`pylint -j1 -rn -sn --enable=all file.py"`

The -j1 option makes pylint use a single threaded reporter which shows the traceback:

```
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/pylint/__main__.py", line 7, in <module>
    pylint.run_pylint()
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/pylint/__init__.py", line 19, in run_pylint
    Run(sys.argv[1:])
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/pylint/lint.py", line 1394, in __init__
    linter.check(args)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/pylint/lint.py", line 801, in check
    self._do_check(files_or_modules)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/pylint/lint.py", line 938, in _do_check
    self.check_astroid_module(ast_node, walker, rawcheckers, tokencheckers)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/pylint/lint.py", line 1018, in check_astroid_module
    walker.walk(ast_node)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/pylint/utils.py", line 1162, in walk
    self.walk(child)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/pylint/utils.py", line 1162, in walk
    self.walk(child)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/pylint/utils.py", line 1162, in walk
    self.walk(child)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/pylint/utils.py", line 1159, in walk
    cb(astroid)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/pylint/checkers/refactoring.py", line 961, in visit_call
    inferred = utils.safe_infer(node.func)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/pylint/checkers/utils.py", line 828, in safe_infer
    value = next(inferit)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/decorators.py", line 92, in wrapped
    res = next(generator)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/decorators.py", line 125, in raise_if_nothing_inferred
    yield next(generator)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/inference.py", line 293, in infer_attribute
    for owner in self.expr.infer(context):
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/util.py", line 148, in limit_inference
    yield from islice(iterator, size)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/context.py", line 107, in cache_generator
    for result in generator:
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/decorators.py", line 125, in raise_if_nothing_inferred
    yield next(generator)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/decorators.py", line 92, in wrapped
    res = next(generator)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/bases.py", line 125, in _infer_stmts
    for inferred in stmt.infer(context=context):
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/util.py", line 148, in limit_inference
    yield from islice(iterator, size)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/context.py", line 107, in cache_generator
    for result in generator:
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/decorators.py", line 125, in raise_if_nothing_inferred
    yield next(generator)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/decorators.py", line 92, in wrapped
    res = next(generator)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/bases.py", line 125, in _infer_stmts
    for inferred in stmt.infer(context=context):
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/util.py", line 148, in limit_inference
    yield from islice(iterator, size)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/context.py", line 107, in cache_generator
    for result in generator:
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/decorators.py", line 125, in raise_if_nothing_inferred
    yield next(generator)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/decorators.py", line 92, in wrapped
    res = next(generator)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/inference.py", line 219, in infer_call
    context=callcontext,
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/scoped_nodes.py", line 1628, in infer_call_result
    yield from returnnode.value.infer(context)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/util.py", line 148, in limit_inference
    yield from islice(iterator, size)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/context.py", line 107, in cache_generator
    for result in generator:
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/decorators.py", line 125, in raise_if_nothing_inferred
    yield next(generator)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/decorators.py", line 92, in wrapped
    res = next(generator)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/inference.py", line 219, in infer_call
    context=callcontext,
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/scoped_nodes.py", line 1628, in infer_call_result
    yield from returnnode.value.infer(context)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/util.py", line 148, in limit_inference
    yield from islice(iterator, size)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/context.py", line 107, in cache_generator
    for result in generator:
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/decorators.py", line 125, in raise_if_nothing_inferred
    yield next(generator)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/decorators.py", line 92, in wrapped
    res = next(generator)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/inference.py", line 211, in infer_call
    for callee in self.func.infer(context):
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/util.py", line 148, in limit_inference
    yield from islice(iterator, size)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/context.py", line 107, in cache_generator
    for result in generator:
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/decorators.py", line 92, in wrapped
    res = next(generator)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/decorators.py", line 125, in raise_if_nothing_inferred
    yield next(generator)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/inference.py", line 293, in infer_attribute
...
    context = contextmod.copy_context(context)
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/context.py", line 146, in copy_context
    return context.clone()
  File "/home/user/venv/pylint-issue-2388/lib/python3.6/site-packages/astroid/context.py", line 96, in clone
    clone = InferenceContext(copy.copy(self.path), inferred=self.inferred)
RecursionError: maximum recursion depth exceeded
```

**Comment by brycepg** [comment]
So it looks like this is just a stack limit issue (is not infinite recursion). The only way to reproduce this issue without pylint is to add extra calls to the call stack:

```python

I think I might merge a few astroid calls to lower the stack size
import astroid

def f():
    return astroid.extract_node("""
    from diofant.simplify import hypersimp, simplify
    def rsolve_hyper():
        b = simplify(g)
        s = hypersimp(b, n)
        s.subs() #@
    """).inferred()

def a(n=5):
    if n > 0:
        return a(n-1)
    else:
        return f()
a()
```

**Comment by brycepg** [comment]
I fixed this issue on a local branch, but running pylint on this package didn't complete even overnight.. Not sure if the fix is worth it -- I think astroid has some perf issues with complex class hierarchies

**Comment by PCManticore** [comment]
We probably need to figure out what's the root cause of this problem. As Bryce mentioned, `pylint` is not even able to complete in a reasonable amount of time. I've tried increasing the recursion limit with `--init-hook="import sys; sys.setrecursionlimit(2000)"` and pylint gets stuck.

**Comment by skirpichev** [comment]
> I've tried increasing the recursion limit with --init-hook="import sys; sys.setrecursionlimit(2000)" and pylint gets stuck.

Is this for whole project tree?

I have tried original example from bugreport (for ``diofant/solvers/recurr.py``) with increased recursion limit (as you suggested, with ``--init-hook``) and the test pass (it's slow, but not hangs forever).

**Comment by PCManticore** [comment]
Just for the original example. I stopped it after a couple of minutes, maybe it would have completed later on.

**Comment by dragoljub** [comment]
I'm seeing the same issue on Python 3.7 and Windows 10 in VS Code. Any ideas how to resolve?

**Comment by tomarharsh** [comment]
Same issue with Python 3.7, Mojave 10.14, VS Code.

**Comment by LuoPengFei** [comment]
Same issue with Python 3.7,  VS Code 1.36.1 (1.36.1),  Mojave 10.14.5

**Comment by karlicoss** [comment]
I'm running into that with the following code snippet:
```
import pandas as pd
df = pd.merge(df)
```
Astroid `1.4.1`; pylint `2.3.1`

**Comment by corndog2000** [comment]
Same error message with Python 3.7.3, Ubuntu 18.04.2, VS Code 1.36.1.

Picture of error: https://imgur.com/a/PTNkteD

**Comment by dragoljub** [comment]
My issue was resolved on Python 3.7.1 after upgrading to these versions:
astroid                            2.2.5
pylint                             2.3.1

**Comment by andreasvc** [comment]
FWIW, I think I encountered the same or a similar bug.

In my case, I tracked down that the problem was with the option `--extension-pkg-whitelist=[...]`. If `numpy` is included in this list, the issue occurs, if I leave it out, it's gone.

**Comment by dragoljub** [comment]
Thanks for the tip. I actually had that setting before and still keep it on. ```--extension-pkg-whitelist=numpy```

However, after upgrading to later versions of python and pylint and astroid I no longer have the issue even with the numpy whitlist setting.

Python 3.7.1
pylint                             2.3.1
astroid                            2.2.5

**Comment by feliwir** [comment]
Any updates on how to get rid of this issue?

**Comment by ashkan-leo** [comment]
Any update on how to solve this issue?

**Comment by PCManticore** [comment]
The original issue is no longer reproduceable with the following versions, installed from `master` branch for both `pylint` and `astroid`:
```
pylint 2.5.0-dev1
astroid 2.4.0
Python 3.8.1 (default, Feb 13 2020, 10:17:07)
[Clang 8.1.0 (clang-802.0.42)]
```

It's also no longer reproduceable with the already released versions of pylint 2.4.4 and astroid 2.3.3.

@skirpichev Please give it a go with the aforementioned versions and let us know if you can still reproduce the issue.

For other folks commenting here, if your issue is not related to `diophant`, please submit a separate one. Thanks!

**Comment by dickreuter** [comment]
The issue is definitely not resolved, neither for 2.4.4, nor for 2.3.1 nor for 2.6

**Comment by ohaibbq** [comment]
I'm also seeing the behavior with `pylint==2.6.0` & `astroid==2.4.2`.

**Comment by dickreuter** [comment]
The strange thing is that it happens intermittently. Rerunning pylint multiple times does not yield the same result. Sometimes it works, sometimes it fails with a recursion error.

**Comment by Yevgnen** [comment]
Facing the same issue with 

```
pylint 2.6.0 python code static checker
├── astroid >=2.4.0,<=2.5
│   ├── lazy-object-proxy >=1.4.0,<1.5.0
│   ├── six >=1.12,<2.0
│   └── wrapt >=1.11,<2.0
├── colorama *
├── isort >=4.2.5,<6
├── mccabe >=0.6,<0.7
└── toml >=0.7.1
```

**Comment by hippo91** [comment]
@dickreuter can you please open a different issue with a reproducer in it please?

**Comment by summer-via** [comment]
I upgrade pylint from 2.5.3 to 2.14.1, then it is solved.
